typings/index.d.ts

```diff
@@ -1,1567 +1,31 @@
-  handlers: Middleware<TContext>[];
-  enterHandlers: Middleware<TContext>[];
-  leaveHandlers: Middleware<TContext>[];
-  ttl?: number;
-}
-
-export class BaseScene<TContext extends SceneContextMessageUpdate> extends Composer<TContext> {
-  constructor(id: string, options?: Partial<BaseSceneOptions<TContext>>)
-
-  id: string;
-
-  options: BaseSceneOptions<TContext>;
-
-  enterHandler: Middleware<TContext>;
-
-  leaveHandler: Middleware<TContext>;
-
-  ttl?: number;
-
-  enter: (...fns: Middleware<TContext>[]) => this;
-
-  leave: (...fns: Middleware<TContext>[]) => this;
-
-  enterMiddleware: () => Middleware<TContext>;
-
-  leaveMiddleware: () => Middleware<TContext>;
-}
-
-export type Scene<TContext extends SceneContextMessageUpdate> = BaseScene<TContext>;
-
-export type StageOptions = SceneContextOptions;
-
-export class Stage<TContext extends SceneContextMessageUpdate> extends Composer<TContext> {
-  constructor(scenes: Scene<TContext>[], options?: Partial<StageOptions>)
-
-  register: (...scenes: Scene<TContext>[]) => this;
-
-  middleware: () => Middleware<TContext>;
-
-  static enter: (sceneId: string, initialState?: object, silent?: boolean) => Middleware<SceneContextMessageUpdate>;
-
-  static reenter: () => Middleware<SceneContextMessageUpdate>;
-
-  static leave: () => Middleware<SceneContextMessageUpdate>;
-}
-
-export function session<TContext extends ContextMessageUpdate>(opts?: Partial<{
-  property: string;
-  store: Map<string, any>;
-  getSessionKey: (ctx: TContext) => string;
-  ttl: number;
-}>): Middleware<TContext>;
-
-export interface TelegrafConstructor extends ComposerConstructor {
-  /**
-   * Initialize new Telegraf app.
-   * @param token Bot token
-   * @param options options
-   * @example
-   * new Telegraf(token, options)
-   */
-  new <TContext extends ContextMessageUpdate>(token: string, options?: TelegrafOptions): Telegraf<TContext>;
-
-  log(logFn?: Function): Middleware<ContextMessageUpdate>;
-}
-
-export interface TOptions {
-
-  /**
-   * Telegram options
-   */
-  telegram?: {
-    /**
-     * https.Agent or http.Agent instance, allows custom proxy, certificate, keep alive, etc.
-     */
-    agent: https.Agent | http.Agent
-
-    /**
-     * Reply via webhook
-     */
-    webhookReply: boolean
+import { BaseScene, Stage } from './stage'
+import { Composer, Middleware } from './composer'
+import { ContextMessageUpdate } from './context'
+import { Extra } from './extra'
+import { Markup } from './markup'
+import { Router } from './router'
+import { session } from './session'
+import { Telegram } from './telegram'
+
+// works around https://github.com/microsoft/TypeScript/issues/37712
+import { Telegraf as _Telegraf } from './telegraf'
+declare class Telegraf<TContext extends ContextMessageUpdate> extends _Telegraf<
+  TContext
+> {}
+
+declare namespace Telegraf {
+  export {
+    BaseScene,
+    Composer,
+    ContextMessageUpdate as Context,
```

Comment:
It's not "not perfect", it's outright incorrect: #920.

I never tried to hide that my changes are breaking. I'm open to exporting it as both `Context` and `ContextMessageUpdate`, but only if @dotcypress allows `telegraf.js` to do the same.

Reply:
You are right: telegraf.js is exporting it as `Context` so ContextMessageUpdate is wrong in the Typescript typings.
https://github.com/telegraf/telegraf/blob/252513a1e80aa444c88e2a33e056cf96023c61d2/telegraf.js#L210

Looks like it has to be breaking the types as it has to do the same as the JavaScript Code does.
As it is `Context` the interface `ContextMessageUpdate` should be renamed to `Context` instead of using `export as` in my opinion.

Using an additional `export {Context as ContextMessageUpdate}` would be nice to not break everything with a simple update